File: 7-event-sourcing/1-java/event-store-service/src/main/java/com/starci/eventstore/AccountProjection.java
package com.starci.eventstore;

import java.util.List;
import java.util.Map;

/**
 * AccountProjection — pure static utility that folds an event log into {@link AccountState}.
 *
 * <p>No side effects, no DB access. Mirrors the TypeScript {@code applyEvent} /
 * {@code replayEvents} pure functions in {@code account-projection.ts}.
 */
public final class AccountProjection {

    private AccountProjection() {}

    /**
     * Apply a single event onto the current state and return the new state.
     * Pure function — does not mutate the input state.
     *
     * @param state      current state
     * @param event      event record to apply
     * @return new {@link AccountState} after applying the event
     */
    public static AccountState applyEvent(AccountState state, EventRecord event) {
        AccountState next = copy(state);
        next.setVersion(event.getVersion());
        next.setLastEventAt(event.getOccurredAt());

        Map<String, Object> payload = event.getPayload();
        String type = (String) payload.get("type");

        switch (type) {
            case "AccountOpened" -> {
                next.setOwner((String) payload.get("owner"));
                next.setBalance(toDouble(payload.get("initialBalance")));
                next.setStatus("open");
                next.setOpenedAt(event.getOccurredAt());
            }
            case "MoneyDeposited" ->
                next.setBalance(state.getBalance() + toDouble(payload.get("amount")));
            case "MoneyWithdrawn" ->
                next.setBalance(state.getBalance() - toDouble(payload.get("amount")));
            case "AccountClosed" ->
                next.setStatus("closed");
            default ->
                throw new IllegalArgumentException("Unknown event type: " + type);
        }
        return next;
    }

    /**
     * Replay a full sorted event list and return the final {@link AccountState}.
     *
     * @param accountId  aggregate ID (used to seed the empty state)
     * @param events     events sorted by version ASC
     * @return           final projected state
     */
    public static AccountState replayEvents(String accountId, List<EventRecord> events) {
        AccountState state = new AccountState(accountId);
        for (EventRecord e : events) {
            state = applyEvent(state, e);
        }
        return state;
    }

    // ─── Helpers ─────────────────────────────────────────────────────────────

    private static AccountState copy(AccountState s) {
        AccountState c = new AccountState();
        c.setAccountId(s.getAccountId());
        c.setOwner(s.getOwner());
        c.setBalance(s.getBalance());
        c.setStatus(s.getStatus());
        c.setVersion(s.getVersion());
        c.setOpenedAt(s.getOpenedAt());
        c.setLastEventAt(s.getLastEventAt());
        return c;
    }

    private static double toDouble(Object v) {
        if (v instanceof Number n) return n.doubleValue();
        return Double.parseDouble(String.valueOf(v));
    }
}

File: 7-event-sourcing/event-store-service/src/account/handlers/account-projection.ts
/**
 * Pure projection function — fold event log → AccountState (read model).
 * Hàm này KHÔNG có side effects, không truy cập DB; chỉ nhận danh sách event và trả về state.
 * (EN: Pure projection function — fold event log → AccountState (read model).
 * This function has NO side effects, no DB access; takes event list and returns state.)
 */
import type {
    AccountDomainEvent,
} from "../events"

/**
 * Trạng thái hiện tại của một Account — được tính lại từ event log.
 * (EN: Current state of an Account — computed by replaying the event log.)
 */
export interface AccountState {
    /** ID tài khoản. (EN: Account ID.) */
    accountId: string
    /** Chủ tài khoản. (EN: Account owner.) */
    owner: string
    /** Số dư hiện tại. (EN: Current balance.) */
    balance: number
    /** Trạng thái: open | closed. (EN: Status: open | closed.) */
    status: "open" | "closed"
    /** Version = số events đã được apply. (EN: Version = number of events applied.) */
    version: number
    /** Thời điểm event đầu tiên (mở TK). (EN: Timestamp of the first event (account opened).) */
    openedAt: Date | null
    /** Thời điểm event cuối cùng được apply. (EN: Timestamp of the last applied event.) */
    lastEventAt: Date | null
}

/**
 * Trạng thái rỗng cho aggregate chưa có event nào.
 * (EN: Empty state for an aggregate with no events yet.)
 */
export function emptyAccountState(accountId: string): AccountState {
    return {
        accountId,
        owner: "",
        balance: 0,
        status: "open",
        version: 0,
        openedAt: null,
        lastEventAt: null,
    }
}

/**
 * Apply một event đơn lẻ vào state hiện tại → trả state mới (immutable fold step).
 * Logic — switch theo eventType, cập nhật field tương ứng.
 * Code — pure function, không mutate tham số đầu vào.
 * (EN: Apply a single event onto current state → return new state (immutable fold step).
 * Logic — switch on eventType, update the corresponding field.
 * Code — pure function, does not mutate input.)
 *
 * @param state - Trạng thái hiện tại. (EN: Current state.)
 * @param event - Event cần apply. (EN: Event to apply.)
 * @param occurredAt - Thời điểm event xảy ra. (EN: When the event occurred.)
 * @param version - Version của event. (EN: Event version.)
 */
export function applyEvent(
    state: AccountState,
    event: AccountDomainEvent,
    occurredAt: Date,
    version: number,
): AccountState {
    const next: AccountState = { ...state, version, lastEventAt: occurredAt }
    switch (event.type) {
    case "AccountOpened":
        return {
            ...next,
            owner: event.owner,
            balance: event.initialBalance,
            status: "open",
            openedAt: occurredAt,
        }
    case "MoneyDeposited":
        return { ...next, balance: state.balance + event.amount }
    case "MoneyWithdrawn":
        return { ...next, balance: state.balance - event.amount }
    case "AccountClosed":
        return { ...next, status: "closed" }
    }
}

/**
 * Replay toàn bộ event log → AccountState cuối cùng.
 * Logic — foldLeft trên mảng event từ version thấp đến cao.
 * Code — `events.reduce(applyEvent, initial)`.
 * (EN: Replay the full event log → final AccountState.
 * Logic — foldLeft over sorted event array low-to-high version.
 * Code — `events.reduce(applyEvent, initial)`.)
 *
 * @param accountId - ID dùng để khởi tạo trạng thái rỗng. (EN: ID used to init empty state.)
 * @param events - Danh sách raw event records đã sort theo version ASC. (EN: Sorted event records ASC version.)
 */
export function replayEvents(
    accountId: string,
    events: Array<{ payload: Record<string, unknown>; occurredAt: Date; version: number }>,
): AccountState {
    let state = emptyAccountState(accountId)
    for (const row of events) {
        state = applyEvent(
            state,
            row.payload as unknown as AccountDomainEvent,
            row.occurredAt,
            row.version,
        )
    }
    return state
}
